### PR TITLE
Brine plume

### PR DESCRIPTION
**Pull request by kshedstrom**
Feel free to squash-merge.

This is a draft of the brine plume parameterization of Nguyen et al., 2009. This shouldn't cause any trouble as is, but requires changes to SIS2 and FMScoupler before it can be turned on. It mixes the brine down from the surface to a depth given by visc%MLD. Further tweaking may occur.

### CODE REVIEW

**Comment by kshedstrom** [comment]
It failed some tests. I tried running them on chinook - the new side failed at:
```
FATAL: NaN detected: ocean_model-ePBL_h_ML

FAIL: tc2.a.symmetric failed at runtime.
```
but the old side did not. On the other hand, the old side failed to build gen_data, then tc4 failed to find the input file:
```
gen_data.F90:99: error: undefined reference to '__netcdf_MOD_nf90_def_var_fill_eightbytereal'
```
The new gen_data wasn't cleaned out by "make clean", but didn't run on the old side due to differing library versions.

**Comment by marshallward** [comment]
Does the netCDF error look like its something on our end?  Or is `nf90_dev_var_fill` truly unavailable?

**Comment by kshedstrom** [comment]
The library does have the missing routine, it's just that the system wasn't pointing to the library. 
```
FCFLAGS = -g -O2 -I/usr/local/pkg/MPI/GCC/11.3.0/OpenMPI/4.1.4/netCDF-Fortran/4.5.4/include
LDFLAGS = 
LIBS = -lnetcdff -lnetcdf 
```
There's no -L/path/to/libdir in LDFLAGS. Also, that path in FCFLAGS is a path on new chinook, not one on old chinook. After manually building gen_data, I'm now getting a different tc4 error (but no tc2 error at all on old chinook):
```
FATAL: fms_io(read_data_3d_new), field x in file ./ocean_hgrid.nc: field size mismatch 1
```

**Comment by marshallward** [comment]
The autoconf ought to determine of -L is required, but perhaps chinook has found some unexpected problem.

Anyway, not the primary issue here.  We can talk privately if needed.

File: src/parameterizations/vertical/MOM_set_viscosity.F90 [hunk 1]
```diff
@@ -1027,22 +1027,24 @@ subroutine set_viscous_BBL(u, v, h, tv, visc, G, GV, US, CS, pbv)
...
-            visc%Ray_u(I,j,k) = visc%Ray_u(I,j,k) + (CS%cdrag*US%L_to_Z*umag_avg(I)) * h_bbl_fr
-          else
-            visc%Ray_v(i,J,k) = visc%Ray_v(i,J,k) + (CS%cdrag*US%L_to_Z*umag_avg(i)) * h_bbl_fr
-          endif
-          h_sum = h_sum + h_at_vel(i,k)
-          if (h_sum >= h_bbl_drag(i)) exit ! The top of this layer is above the drag zone.
-        enddo
-        ! Do not enhance the near-bottom viscosity in this case.
-        Kv_bbl = CS%Kv_BBL_min
+      if (CS%body_force_drag) then
```
Comment (Hallberg-NOAA) [inline src/parameterizations/vertical/MOM_set_viscosity.F90:1030]: Is there a reason why these two logical conditions should be split into separate if tests?

  (At the moment I am being extra-vigilant for any changes that are likely to create conflicts with what I am doing with the non-Boussinesq conversion, and it seems like this change in MOM_set_viscosity might be one of them.)

**Comment by kshedstrom** [comment]
This is one of those cases Marshall and I were chatting about on slack: in debug mode, the model will crash if h_bbl_drag is not well defined and CS%body_force_drag is false.

**Review by Hallberg-NOAA — APPROVED** [review]
We have discussed this PR extensively, and I think that it can be safely merged into dev/gfdl.  The new options in this PR should perhaps be considered experimental, in that there are some scenarios where issues might emerge, such as when there is wetting and drying or when the boundary layer reaches the seafloor leading to large salt fluxes being applied to thin layers, or when convective instabilities are created. However, these emerging issues can be debugged and corrected as they arise.  Moreover, this new option can be completely disabled via the master switch DO_BRINE_PLUME, so even if problems do emerge, they should not impact any other runs.

**Comment by marshallward** [comment]
@kshedstrom I am about to test and merge this, but I recall you saying there was an issue with some GCC compilers.  Is that still the case?

**Comment by kshedstrom** [comment]
No, my issue is with FMS 2023.01.01.

**Comment by marshallward** [comment]
OK, great.  We'll talk offline about the GFortran FMs issue.

**Comment by marshallward** [comment]
Gaea regression: https://gitlab.gfdl.noaa.gov/ogrp/MOM6/-/pipelines/20027 :heavy_check_mark: :yellow_circle:

**Pull request merged**